src/portable/synopsys/dwc2/dcd_dwc2.c

```diff
@@ -630,6 +630,27 @@ void dcd_sof_enable(uint8_t rhport, bool en)
 /* DCD Endpoint port
  *------------------------------------------------------------------*/
 
+// Check is IN/OUT endpoint is avaliable before opening it
+static bool dcd_edpt_check_if_avaliable(uint8_t rhport, uint8_t direction)
```

Comment:
The logic of function `dcd_edpt_check_if_avaliable` is next: verify the in_ep or out_ep >= 1 and if yes, then decrease both of them. 
So, logically, it is better to have the function with clean logic, such as comparing one value with the other and make an assert based on the result of this function. 

So, the idea should be: 
1. Amount of opened IN EP should be less than EP_MAX_SIZE - 1 (only for ESP32S3)
2. Summ of opened IN EP and OUT EP should be  less than EP_MAX_SIZE

In this case, seems that it is better:
1. To have a counter for IN EP and OUT EP
2. Verify possibility to open new EP (with assert)
3. Increase counter after successful opening EP
4. Decrease counter after closing EP 

IMO, it will be a bit better.

Reply:
Thanks for the review. Indeed I did not take into account closing the endpoints.
Will take a look at what you are suggesting.